pyccel/parser/semantic.py

```diff
@@ -3756,17 +3866,33 @@ def _visit_Return(self, expr):
             expr  = Return(results)
         return expr
 
-    def _visit_FunctionDef(self, expr):
+    def _visit_FunctionDef(self, expr, annotate=False, function_call=None):
+
+        if not annotate:
+            self.insert_function(expr)
+            return EmptyNode()
```

Comment:
this will cause a problem, we need the function to be present in the `functions` dict, so that `self.scope.find('functions', name)` will return something

Reply:
Yes. But could this not be done with a loop in `_visit_Module` so you don't need to juggle this extra parameter?

Reply:
Yes, it's doable, but then I will be traversing the tree twice, even though it is not costly in this case, but I try to avoid it if not necessary

Reply:
Is it still traversing if you just do insert:
```
for f in mod.funcs:
    self.insert_function(f)
```
? I prefer to avoid changing the documented prototype if not necessary too  so we have to find a balance here depending on what should be best practice.

Reply:
Actually my suggestion won't work. This is not the same thing. Why is the `insert_function` not called at the end of `_visit_FunctionDef`? Wouldn't that fix the problem?

Reply:
we have one at the end of the `_visit_FunctionDef`  but only used if the function is annotated

Reply:
the `annotate` tells us if we should annotate it or not, not that it is already annotated, so your suggested fix by using the attribute `is_annotated` won't work

Reply:
I have added a doc string in [783adc6](https://github.com/pyccel/pyccel/pull/1728/commits/783adc6ae40bfab4c082e86cc297bec661820f73)

Reply:
In what circumstances would we visit a `FunctionDef` without annotating it?

Reply:
in the first pass, e.g
```python
def f():
    pass
f()
```
when we start annotating the Module, the function f will not be annotated in the first pass, it will only be 
annotated in the function call f()`

Reply:
Huh, I didn't realise that we visited functions inside `CodeBlock` objects. I assumed that we put them in the Module/FunctionDef in the syntactic stage. If we did this then we would simply need a loop to do `insert_function`. Without it I agree that we need this extra parameter but I find this double visitation less clean

Reply:
before we had the assumption of inserting functions in the scope only if they are annotated, now it is different, we insert them even if they are not annotated, which means that we can insert them in the syntactic stage, but then we probably need to the same with imports, classes also, so I'm not sure. we can do it in another PR, the changes will be easy for functions, we just delete the flag and where it is passed, which is maybe 5 lines in the semantics file

Reply:
Ok. Then we can leave this simplification for a separate PR :+1: Please create an issue so we don't forget to do this

Reply:
done https://github.com/pyccel/pyccel/issues/1768

Reply:
Thanks :)